### PR TITLE
EGA basics

### PR DESCRIPTION
**Pull request by Macroz**
# Checklist for author

For #2466.

Everything that is implemented so far for early comments and feedback. I think we need a set of features required for MVP demo. I implemented some of the UI too but it's not required in the first demo (where Approver Bot needs special features that are not here yet).

Remove items that aren't applicable, check items that are done.

## Reviewability
- [x] Link to issue
- [ ] Note if related change in rems-deploy repo TBD
- [ ] Consider adding screenshots for ease of review TBD

## Backwards compatibility
- [x] API is backwards compatible or completely new
- [x] Config is backwards compatible

## Documentation
- [ ] Update changelog if necessary TBD
- [ ] API is documented and shows up in Swagger UI
- [ ] Components are added to guide page
- [ ] Update docs/ (if applicable) TBD
- [ ] Update manual/ (if applicable) TBD
- [ ] ADR for major architectural decisions or experiments TBD
- [x] New config options in config-defaults.edn

## Testing
- [x] Complex logic is unit tested
- [x] Valuable features are integration / browser / acceptance tested automatically (more TBD)

## Follow-up
- [ ] New tasks are created for pending or remaining tasks TBD

### CODE REVIEW

File: src/clj/rems/db/entitlements.clj [hunk 1]
```diff
@@ -128,7 +145,10 @@
                            :resource resource-id
                            :revokedby actor
                            :end end})
-    (add-to-outbox! :remove (db/get-entitlements {:application application-id :user user-id :resource resource-id}))))
+    (let [entitlements (db/get-entitlements {:application application-id :user user-id :resource resource-id})]
+      (add-to-outbox! :remove :basic entitlements nil)
```
Comment (opqdonut) [inline src/clj/rems/db/entitlements.clj:149]: could maybe add a TODO about deprecating the :basic functionality

File: src/clj/rems/db/user_secrets.clj [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        coerce-user-secrets)))
+
+(defn get-user-secrets [user]
+  (when user
+    (json->secrets (:secrets (db/get-user-secrets {:user user})))))
+
+(defn update-user-secrets! [user new-secrets]
+  (assert user "User missing!")
+  (let [old-secrets (get-user-secrets user)
+        validated (try (validate-user-secrets new-secrets) (catch Exception _e nil))]
```
Comment (opqdonut) [inline src/clj/rems/db/user_secrets.clj:34]: hmm shouldn't you validate `(merge old-secrets new-secrets)` ?

File: src/clj/rems/db/user_secrets.clj [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    (if (= (set (keys validated))
+           (set (keys new-secrets)))
```
Comment (opqdonut) [inline src/clj/rems/db/user_secrets.clj:36]: is this keyset check needed? doesn't schema catch it for us?

File: src/clj/rems/ega.clj [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      :body
+      json/parse-string))
+
+(defn api-key-list
+  "List the API-Keys available.
+
+  `:access-token` - the access token
+  `:config`       - configuration of the EGA integration
+
+  NB: The token itself cannot be returned anymore."
```
Comment (opqdonut) [inline src/clj/rems/ega.clj:66]: I'm not sure what this means

File: src/clj/rems/ega.clj [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  [{:keys [access-token config]}]
+  (-> (http/get (str (:permission-server-url config) "/api_key")
+                {:oauth-token access-token
+                 :socket-timeout 2500
+                 :conn-timeout 2500})
+      :body
+      json/parse-string))
+
+(defn read-permissions
```
Comment (opqdonut) [inline src/clj/rems/ega.clj:75]: Inconsistency: This and the next function are prefixed `read-` even though it's not in the api path. Other GET requests aren't prefixed `read-`.

File: src/clj/rems/ega.clj [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  "Lists the users with permission to the dataset.
+
+  `:api-key`    - valid API-Key of the person asking
+  `:dataset-id` - id of the dataset
+  `:config`     - configuration of the EGA integration"
+  [{:keys [api-key dataset-id config]}]
+  (-> (http/get (str (:permission-server-url config) "/datasets/" dataset-id "/users")
+                {:headers {"authorization" (str "api-key " api-key)}
+                 :socket-timeout 2500
+                 :conn-timeout 2500})
```
Comment (opqdonut) [inline src/clj/rems/ega.clj:117]: should these repeating timeouts be factored out?

File: src/clj/rems/ega.clj [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      :body
+      json/parse-string))
```
Comment (opqdonut) [inline src/clj/rems/ega.clj:119]: I think you can just pass `:as :json` to `http/get`.

File: src/cljc/rems/common/catalogue_util.cljc [hunk 8]
```diff
@@ -1,9 +1,33 @@
...
 
-(defn urn? [resid]
+(defn- urn? [resid]
   (and resid (str/starts-with? resid "urn:")))
 
-(defn urn-catalogue-item-link [{:keys [resid]} {:keys [urn-organization]}]
+(defn- urn-catalogue-item-url [resid {:keys [urn-organization]}]
   (when (urn? resid)
     (str (or urn-organization "http://urn.fi/") resid)))
+
```
Comment (opqdonut) [inline src/cljc/rems/common/catalogue_util.cljc:11]: the EGA url stuff could be split out to a separate PR and just merged

File: src/clj/rems/db/user_settings.clj [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
   {:language (:default-language env)
    :notification-email nil})
 
 ;; TODO should this be in schema-base?
 (s/defschema UserSettings
-  {:language s/Keyword
-   :notification-email (s/maybe s/Str)})
+  {(s/optional-key :language) s/Keyword
+   (s/optional-key :notification-email) (s/maybe s/Str)
+   (s/optional-key :ega) {:api-key-expiration-date DateTime}})
```
Comment (opqdonut) [inline src/clj/rems/db/user_settings.clj:19]: This refactor could be split out to its own PR and discussed separately. I'm slightly concerned that we won't guarantee the presence of any of these keys, even though the UI expects `:language` and `:notification-email`.

**Review by opqdonut — COMMENTED** [review]
first round of comments

File: src/clj/rems/ega.clj (hunk 4 above)
Comment (Macroz) [reply src/clj/rems/ega.clj:66]: I mean that the API-key cannot be fetched from the API, it is only seen once when generating it.

File: src/clj/rems/db/user_secrets.clj (hunk 3 above)
Comment (Macroz) [reply src/clj/rems/db/user_secrets.clj:36]: Yeah I think it's not needed, it is legacy from user settings.

File: src/clj/rems/db/user_secrets.clj (hunk 2 above)
Comment (Macroz) [reply src/clj/rems/db/user_secrets.clj:34]: :+1: I'll rewrite that part.

File: src/clj/rems/db/entitlements.clj (hunk 1 above)
Comment (Macroz) [reply src/clj/rems/db/entitlements.clj:149]: :+1: It's in my notes that I'll probably put in the issue or actually for this create a new one.

File: src/clj/rems/ega.clj (hunk 5 above)
Comment (opqdonut) [reply src/clj/rems/ega.clj:75]: I like naming the functions according to the API endpoints, and I think some sort of get- / read- / fetch- prefix might work here, but it needs to be applied consistently.

File: src/clj/rems/ega.clj (hunk 4 above)
Comment (Macroz) [reply src/clj/rems/ega.clj:66]: How about? `NB: The actual API-Key is not returned by this call. It is only returned once in `api-key-generate`.`

Comment (Macroz) [reply src/clj/rems/ega.clj:66]: Could also remove completely, it's not important.

Comment (opqdonut) [reply src/clj/rems/ega.clj:66]: Either of those is fine by me!

File: src/clj/rems/ega.clj (hunk 5 above)
Comment (Macroz) [reply src/clj/rems/ega.clj:75]: You are right, I think I will prefix all the methods with the request-method so that they become unique. Otherwise GET permissions and DELETE permissions would be the same.

Comment (Macroz) [reply src/clj/rems/ega.clj:75]: They can also all be changed into private?

File: src/clj/rems/ega.clj (hunk 6 above)
Comment (Macroz) [reply src/clj/rems/ega.clj:117]: Probably makes sense. I'm not sure if they are actually needed at all or what they should be. The entitlement post used to have these but I'm not sure if they make sense?

Comment (opqdonut) [reply src/clj/rems/ega.clj:117]: see also timeouts in `rems.event-notification`

File: src/clj/rems/db/user_settings.clj (hunk 9 above)
Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: The API did not guarantee them before either?

Comment (opqdonut) [reply src/clj/rems/db/user_settings.clj:19]: No, the keys are required in the API response, see
- how rems.api.user-settings usess the user-settings/UserSettings schema
- https://rems-dev.rahtiapp.fi/swagger-ui/index.html#/user-settings/get_api_user_settings

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: How is it possible? It used to reference `PartialUserSettings` schema where they were optional.

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: Ah, I misread. Only PUT referenced them.

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: Since these values come from the `default-settings` if they don't exist we could force the presence of the keys like before.

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: In here the original problem is doing it differently, and skipping separate read and write models. Cleaning up in a backwards compatible way is difficult. But I could introduce separate read & write models.

Comment (opqdonut) [reply src/clj/rems/db/user_settings.clj:19]: Yeah that's why I wanted to split this out into a separate PR, we need to do this carefully. I like the schema simplification. Maybe we can have something like `(defn UserSettingsInApi (assoc UserSettings :language s/Keyword :notification-email s/Str))` instead of the `PartialUserSettings` hack.

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: We can have separate DB and API schemas as the others?

Comment (opqdonut) [reply src/clj/rems/db/user_settings.clj:19]: Yeah can just do full decoupling

Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: I'll do that and split the small parts out.

File: src/clj/rems/ega.clj (hunk 6 above)
Comment (Macroz) [reply src/clj/rems/ega.clj]: I think we'll have to see about EGA what they should eventually be. I've extracted them into local constants for now.

File: src/clj/rems/db/user_settings.clj (hunk 9 above)
Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: This PR is now updated with what I think is the separate db, read and write models with backwards compatibility. I have to look at it with fresh eyes still. I'll harvest them to another PR next.

File: src/cljc/rems/common/catalogue_util.cljc (hunk 8 above)
Comment (Macroz) [reply src/cljc/rems/common/catalogue_util.cljc:11]: https://github.com/CSCfi/rems/pull/2605

File: src/clj/rems/db/user_settings.clj (hunk 9 above)
Comment (Macroz) [reply src/clj/rems/db/user_settings.clj:19]: https://github.com/CSCfi/rems/pull/2606

File: src/clj/rems/db/user_secrets.clj [hunk 10]
```diff
@@ -31,9 +31,9 @@
 (defn update-user-secrets! [user new-secrets]
   (assert user "User missing!")
   (let [old-secrets (get-user-secrets user)
-        validated (try (validate-user-secrets new-secrets) (catch Exception _e nil))]
-    (if (= (set (keys validated))
-           (set (keys new-secrets)))
+        validated (try (validate-user-secrets (merge old-secrets new-secrets))
+                       (catch Exception _e nil))]
```
Comment (opqdonut) [inline src/clj/rems/db/user_secrets.clj:35]: should probably have some sort of `{:success false}` logic here (just like in user-settings) instead of failing silently.

Comment (Macroz) [reply src/clj/rems/db/user_secrets.clj:35]: The next line is the `if` that goes to `{:success false}` if we got the `nil` here?

Comment (opqdonut) [reply src/clj/rems/db/user_secrets.clj:35]: Oh right, I'm blind.

**Comment by Macroz** [comment]
I'm closing this draft now. Towards production code in https://github.com/CSCfi/rems/pull/2637.
